node/src/components/small_network/task_manager.rs

```diff
@@ -0,0 +1,231 @@
+
+impl ShutdownReceiver {
+    /// Waits for the shutdown signal.
+    ///
+    /// The returned future is safe to cancel.
+    pub async fn wait_for_shutdown(&mut self) {
+        match self.0.changed().await {
+            Ok(()) => {
+                // We are using the dropping of the channel as a signal, not the channel itself!
+                error!("received value on shutdown channel, this should never happen")
+            }
+            Err(_recv_err) => {
+                // All good.
+            }
+        }
+    }
+}
+
+/// The outcome of awaiting a task for shutdown.
+#[derive(Debug)]
+enum ShutdownOutcome {
+    /// The task was shut down completely.
+    Ok,
+    /// Could not join the task, it either panicked or was explicitly cancelled.
+    JoinFailure(JoinError),
+    /// Join timeout.
+    ///
+    /// The task is still running in the background, the join handle has been lost.
+    TimedOut,
+}
+
+/// Manager for tasks.
+///
+/// Can be used to signal a group of tasks to shutdown and wait for their completion, optionally
+/// cancelling execution. All operations are "best effort", won't panic even on poisoned locks.
+///
+/// Dropping a task manager will instruct all tasks to shutdown, but the `drop` will not await their
+/// termination.
+#[derive(DataSize, Debug)]
+pub struct TaskManager {
+    /// A counter for joined tasks.
+    ///
+    /// Incremented to provide unique IDs for tasks.
+    task_counter: usize,
+    /// Join handles for all spawned tasks, mapped from task ID to join handle.
+    join_handles: Arc<Mutex<HashMap<usize, JoinHandle<()>>>>,
+    /// Sender for the signal to shutdown.
+    #[data_size(skip)]
+    shutdown_sender: watch::Sender<()>,
+    /// Receiver for the signal to shutdown.
+    shutdown_receiver: ShutdownReceiver,
+}
+
+impl Default for TaskManager {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TaskManager {
+    /// Creates a new task manager.
+    pub fn new() -> Self {
+        let (shutdown_sender, shutdown_watch_receiver) = watch::channel(());
+
+        Self {
+            task_counter: 0,
+            join_handles: Default::default(),
+            shutdown_sender,
+            shutdown_receiver: ShutdownReceiver(shutdown_watch_receiver),
+        }
+    }
+
+    /// Spawns a new tasks.
+    ///
+    /// Tasks are expected to honor received shutdown signals from the manager, that is they need to
+    /// shutdown once they receive a value on the `Receiver`. No return values are supported for
+    /// spawned tasks.
+    ///
+    /// If the internal log has been poisoned, outputs a warning, but otherwise keeps on going.
+    pub fn spawn<F, G>(&mut self, gen: F)
+    where
+        F: FnOnce(ShutdownReceiver) -> G,
+        G: Future<Output = ()> + Send + 'static,
+    {
+        let task_id = self.task_counter;
+        self.task_counter += 1;
+
+        let join_handles = self.join_handles.clone();
```

Comment:
If we close over a strong `Arc` in a closure/future, it can keep the item alive longer than intended. This can be a source of resource leaks. Instead, let's prefer `Weak` here, with `Arc::downgrade()`, and test with `weak_join_handles.upgrade()` in the future itself.